File: modules/allnewmts-lua/android/src/main/java/com/allnewmts/lua/AllNewMTSLuaModule.kt
package com.allnewmts.lua

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class AllNewMTSLuaModule : Module() {
  private var runtime = 0L

  init {
    System.loadLibrary("allnewmts_lua")
  }

  override fun definition() = ModuleDefinition {
    Name("AllNewMTSLua")

    Function("create") {
      if (runtime != 0L) nativeDestroy(runtime)
      runtime = nativeCreate()
      runtime != 0L
    }

    Function("evaluate") { source: String ->
      nativeEvaluate(runtime, source)
    }

    Function("destroy") {
      if (runtime != 0L) nativeDestroy(runtime)
      runtime = 0L
    }
  }

  private external fun nativeCreate(): Long
  private external fun nativeEvaluate(runtime: Long, source: String): String
  private external fun nativeDestroy(runtime: Long)
}

File: modules/allnewmts-lua/ios/AllNewMTSLuaModule.swift
import ExpoModulesCore

public final class AllNewMTSLuaModule: Module {
  private let adapter = AllNewMTSLuaAdapter()

  public func definition() -> ModuleDefinition {
    Name("AllNewMTSLua")

    Function("create") { () -> Bool in
      adapter.create()
    }

    Function("evaluate") { (source: String) throws -> String in
      var error: NSError?
      guard let value = adapter.evaluate(source, error: &error) else {
        throw error ?? NSError(domain: "AllNewMTSLua", code: 1)
      }
      return value
    }

    Function("destroy") { () in
      adapter.destroy()
    }
  }
}
